rfcs/0074-tvmscript-unified-printer.md

```diff
@@ -0,0 +1,475 @@
+├── tir # TIR dialect 
+│   ├── expr.cc
+│   ├── stmt.cc
+│   └── ...
+└── relax # Hypothetical Relax dialect (not part of our RFC)
+    └── ...
+```
+
+**Goal 2 [Third-Party IRs in Multi-Stage Compilation].** Modularize and
+infrastructuralize the printer to support more future IRs or third-party IRs at
+any level with maintainability, for example, IRs at lower-level than TIR, or
+Relax VM executable.
+
+The current TVMScript printer is tightly coupled with TIR by being a subclass
+of TIR-specific functors
+([link](https://github.com/apache/tvm/blob/main/src/printer/tvmscript_printer.cc#L129)).
+This design isn’t scalable when we want to support more IRs. More importantly,
+it’s impossible for the current approach to support third-party IR bteing
+registered in a dynamic library.
+
+**Goal 3 [Reproducibility and Error Reporting].** Expand reproducibility and
+flexible rendering of diagnostic messages during any level of IR
+transformation.
+
+For example, the following snippet runs and produces an error.
+
+```py
+import tvm
+
+@T.prim_func
+def func_a(A: T.Buffer[(1,), "int32"]):
+    A[0] = 0
+
+@T.prim_func
+def func_b(A: T.Buffer[(8,), "int32"]):
+    A[0] = 0
+
+tvm.ir.assert_structural_equal(func_a, func_b)
+```
+
+The current error message indicates what the difference was, but not where it
+occurred.  This can sometimes be inferred from a stack trace, but becomes
+increasingly difficult with larger IR graphs.
+
+```
+ValueError: StructuralEqual check failed, caused by lhs:
+1
+and rhs:
+8
+```
+
+TUNIP should enable individual utilities and IR passes to have error messages
+directing the user to exact locations in the IR representation.
+
+```
+ValueError: StructuralEqual check failed, first delta highlighted below
+
+@T.prim_func
+def func_a(A: T.Buffer[(1,), "int32"]) -> None:
+                       ^^^^
+    A[0] = 0
+
+@T.prim_func
+def func_b(A: T.Buffer[(8,), "int32"]) -> None:
+                       ^^^^
+    A[0] = 0
+```
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This section introduces the design philosophy of the printer, and demonstrates
+the proposed user-facing APIs where users means IR developers.
+
+## Two-Stage Translation
+
+Traditionally in TVM stack, printing is a single-stage process. The printer
+assumes certain syntax of the target language, and therefore, so far there are
+3 different printers all for TIR: ReprPrinter, TIRTextPrinter,
+TVMScriptPrinter.
+
+We extend the idea of the existing Doc class at
+[src/printer/doc.h#L67](https://github.com/apache/tvm/blob/main/src/printer/doc.h#L67)
+to allow better consistency and scalability. An IR, which could
+be TIR, Relax or any other ones developed by third-party vendors, is first
+translated to an intermediate Doc node tree, and then the Doc tree is mapped to a target
+language, for example, Python, C++ IRBuilder API, or Rust.
+
+**Stage 1 [TVM IR => Doc]**. On the first stage, the printer needs to take care
```

Comment:
just to spell it out, what's the rationale for translating into Doc tree if we have to then do another step of translating the Doc tree into a target frontend language? will the Doc tree be reused e.g. in the parser? are there common operations we can handle in translating into the Doc tree?

Reply:
okay but i'm not quite sure i follow how we are splitting up the printing operations here when we are considering supporting multiple frontend languages. is it like:
1. Doc tree represents a sort of AST that captures things like scope nesting which may affect indent, without specifying how to serialize ForDoc
2. Doc tree -> language translation step prints ForNode in the language proper?

how do we handle e.g. Python indentation when translating from Doc -> Python? is it assumed that the IRModule -> doc translation always sets indentation that's compatible with Python?

Reply:
I added a section for the rationale of translating into Doc tree in RFC. Briefly speaking, two-stage printing is more scalable if printer needs to support more than one IR, because it separates the logic to handle IR semantics with the logic to handle target language syntax. 

<details>
  <summary>Code Example</summary>
  
To print buffer region (things like `A[1:2, 3, 4]`) in the current TVMScript printer,
```cpp
Doc TVMScriptPrinter::PrintBufferRegion(const BufferRegionNode* op) {
  Doc doc;
  if (op->region.size() == 0) {
    doc << Print(op->buffer) << "[()]";
  } else {
    doc << Print(op->buffer) << "[";
    for (size_t i = 0; i < op->region.size(); ++i) {
      if (i != 0) doc << ", ";
      const auto& range = op->region[i];
      if (!is_one(range->extent)) {
        doc << Print(range->min) << " : " << Print(ana_.Simplify(range->min + range->extent));
      } else {
        doc << Print(range->min);
      }
    }
    doc << "]";
  }
  return doc;
}
```

while in the unified printer with two-stage printing

```cpp
ExprDoc PrintBufferRegion(tir::BufferRegion buffer_region, IRDocsifier p) {
  Array<Doc> indices;

  for (const Range& range : buffer_region->region) {
    if (tir::is_one(range->extent)) {
      indices.push_back(p->AsExprDoc(range->min));
    } else {
      indices.push_back(p->AsExprDoc(range));
    }
  }

  return p->AsExprDoc(buffer_region->buffer)->Index(indices);
}
```

</details>

Reply:
@areusch To answer your questions,

> is it assumed that the IRModule -> doc translation always sets indentation that's compatible with Python?
The IR -> Doc translation doesn't know about indentation. It just sets the right values to Doc without knowing any detail on target language syntax. For example, ForDoc is defined as
```cpp
class ForDocNode : public StmtDocNode {
 public:
  ExprDoc lhs{nullptr};
  ExprDoc rhs{nullptr};
  Array<StmtDoc> body;
}
```

In the Doc -> language translation, indentation is added when it processes the `ForDoc->body`.